hermes-client/src/main/java/pl/allegro/tech/hermes/client/ReactiveHermesClient.java

```diff
@@ -0,0 +1,324 @@
+package pl.allegro.tech.hermes.client;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Scheduler;
+import reactor.core.scheduler.Schedulers;
+import reactor.util.retry.Retry;
+
+import java.net.URI;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Predicate;
+
+import static pl.allegro.tech.hermes.client.HermesMessage.hermesMessage;
+import static reactor.core.Exceptions.isRetryExhausted;
+
+
+public class ReactiveHermesClient {
+    private static final Logger logger = LoggerFactory.getLogger(ReactiveHermesClient.class);
+    private static final String RETRY_CONTEXT_KEY = "hermes-retry-context-key";
+
+    private final ReactiveHermesSender sender;
+    private final String uri;
+    private final Map<String, String> defaultHeaders;
+    private final AtomicInteger currentlySending = new AtomicInteger(0);
+    private final int maxRetries;
+    private final Predicate<HermesResponse> retryCondition;
+    private final Duration retrySleep;
+    private final Duration maxRetrySleep;
+    private final Scheduler scheduler;
+    private volatile boolean shutdown = false;
+    private final List<MessageDeliveryListener> messageDeliveryListeners = new ArrayList<>();
```

Comment:
Why not make those a part of the constructor?

Reply:
Good idea.